### PR TITLE
(RE-15124) Initial implementation

### PR DESCRIPTION
**Pull request by yachub**
Creates a DNS plugin for GCP.

Depends on https://github.com/puppetlabs/vmpooler/pull/551

### CODE REVIEW

File: .github_changelog_generator
Comment (genebean) [inline .github_changelog_generator:1]: Please add new line to end of file

File: .jrubyrc
Comment (genebean) [inline .jrubyrc:1]: Please add new line to end of file

File: .rubocop.yml
Comment (genebean) [inline .rubocop.yml:1]: Please add new line to end of file

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+## [Unreleased]
+
+## [0.1.0] - 2023-01-30
```
Comment (genebean) [inline CHANGELOG.md:3]: Update changelog to real date when taking out of draft

File: Gemfile [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+end
+
+# Evaluate ~/.gemfile if it exists
+if File.exists?(File.join(Dir.home, '.gemfile'))
+  instance_eval(File.read(File.join(Dir.home, '.gemfile')))
+end
+
+group :development do
+  gem 'vmpooler', git: 'https://github.com/puppetlabs/vmpooler.git', branch: 'vmpooler-dns-gcp'
+end
```
Comment (genebean) [inline Gemfile:17]: Update before taking out of draft and add a new line at the end of the file

File: lib/vmpooler/dns/gcp.rb [hunk 3]
```diff
@@ -0,0 +1,142 @@
+# frozen_string_literal: true
+
+require 'googleauth'
+require 'google/cloud/dns'
+require 'vmpooler/dns/base'
+
+module Vmpooler
+  class PoolManager
+    class Dns
+      # This class represent a DNS plugin to CRUD resources in google clouddns.
```
Comment (genebean) [inline lib/vmpooler/dns/gcp.rb:10]:
```suggestion
      # This class represent a DNS plugin to CRUD resources in Google CloudDNS.
```

File: README.md [hunk 4]
```diff
@@ -1 +1,49 @@
...
+- [vmpooler-dns-gcp](#vmpooler-dns-gcp)
+  - [Requirements](#requirements)
+  - [Usage](#usage)
+  - [Update the Gemfile Lock](#update-the-gemfile-lock)
+  - [Releasing](#releasing)
+  - [License](#license)
+
+## Requirements
+
+1. A Google Cloud Project with the [Cloud DNS](https://cloud.google.com/dns/) enabled.
```
Comment (genebean) [inline README.md:12]:
```suggestion
1. A Google Cloud Project with the [CloudDNS](https://cloud.google.com/dns/) enabled.
```
I am fairly sure there is no space...

File: lib/vmpooler/dns/gcp.rb [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+        def initialize(config, logger, metrics, redis_connection_pool, name, options)
+          super(config, logger, metrics, redis_connection_pool, name, options)
+
+          task_limit = global_config[:config].nil? || global_config[:config]['task_limit'].nil? ? 10 : global_config[:config]['task_limit'].to_i
+
+          default_connpool_size = [provided_pools.count, task_limit, 2].max
+          # connpool_size = provider_config['connection_pool_size'].nil? ? default_connpool_size : provider_config['connection_pool_size'].to_i
+          connpool_timeout = 60
+          # logger.log('d', "[#{name}] ConnPool - Creating a connection pool of size #{connpool_size} with timeout #{connpool_timeout}")
```
Comment (genebean) [inline lib/vmpooler/dns/gcp.rb:23]: What's with the commented out bits?

File: lib/vmpooler/dns/gcp.rb [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+          @connection_pool = Vmpooler::PoolManager::GenericConnectionPool.new(
+            metrics: metrics,
+            connpool_type: 'dns_connection_pool',
+            connpool_provider: name,
+            size: default_connpool_size,
+            # size: connpool_size,
+            timeout: connpool_timeout
+          ) do
+            logger.log('d', "[#{name}] Connection Pool - Creating a connection object")
+            # Need to wrap the vSphere connection object in another object. The generic connection pooler will preserve
```
Comment (genebean) [inline lib/vmpooler/dns/gcp.rb:34]: The vSphere reference here seems out of place to me

File: spec/gcp_helper.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  :scope, :retries, :timeout
+) do
+  def zone
+    MockDnsZone.new
+  end
+end
+
+def mock_Google_Cloud_Dns_Project_Connection(options = {})
+  MockGoogleCloudDnsProjectConnection.new()
+end
```
Comment (genebean) [inline spec/gcp_helper.rb:26]: Please add new line to end of file

File: spec/spec_helper.rb
Comment (genebean) [inline spec/spec_helper.rb:1]: Please add new line to end of file

File: spec/unit/dns/gcp_spec.rb
Comment (genebean) [inline spec/unit/dns/gcp_spec.rb:1]: Please add new line to end of file

File: vmpooler-dns-gcp.gemspec [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+lib = File.expand_path('../lib', __FILE__)
+$LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
+require 'vmpooler-dns-gcp/version'
+
+Gem::Specification.new do |spec|
+  spec.name    = "vmpooler-dns-gcp"
+  spec.version = VmpoolerDnsGcp::VERSION
+  spec.authors = ["Puppet by Perforce"]
+  spec.email   = ["support@puppet.com"]
```
Comment (genebean) [inline vmpooler-dns-gcp.gemspec:11]: Please fix this in all the places

File: vmpooler-dns-gcp.gemspec [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  spec.metadata["changelog_uri"] = "https://github.com/puppetlabs/vmpooler-dns-gcp/blob/main/CHANGELOG.md"
+
+  # Specify which files should be added to the gem when it is released.
+  spec.files = Dir[ "lib/**/*" ]
+  spec.require_paths = ["lib"]
+  spec.add_dependency "googleauth", ">= 0.16.2", "< 1.3.0"
+  spec.add_dependency "google-cloud-dns", "~> 0.35.1"
+  # spec.add_dependency 'vmpooler', '>= 1.3.0', '~> 2.3'
+
+  # Testing dependencies
```
Comment (genebean) [inline vmpooler-dns-gcp.gemspec:28]: Fix before coming out of draft

**Review by genebean — CHANGES_REQUESTED** [review]

File: README.md (hunk 4 above)
Comment (yachub) [reply README.md:12]: To my surprise the link to the docs on that line actually does have a space between them

File: lib/vmpooler/dns/gcp.rb (hunk 5 above)
Comment (yachub) [reply lib/vmpooler/dns/gcp.rb:23]: Leftover copy/paste, I think if it needs customized outside of the computed size then that can be added as an additional feature.

File: CHANGELOG.md (hunk 1 above)
Comment (yachub) [reply CHANGELOG.md:3]: I reset the changelog, so naturally will have to update it in a release prep PR post merge as expected.

File: vmpooler-dns-gcp.gemspec (hunk 8 above)
Comment (genebean) [reply vmpooler-dns-gcp.gemspec:11]: Fine by me.

File: .jrubyrc
Comment (genebean) [inline .jrubyrc:1]: @yachub should this file be added to the other vmpooler repos?

File: .rubocop.yml
Comment (genebean) [inline .rubocop.yml:1]: Please add a new line to the end of the file

File: Gemfile.lock [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+      redis (~> 4.1)
+      sinatra (~> 2.0)
+      spicy-proton (~> 2.1)
+      statsd-ruby (~> 1.4)
+    webrick (1.8.1)
+    zonefile (1.06)
+
+PLATFORMS
+  universal-java-1.8
+  universal-java-11
```
Comment (genebean) [inline Gemfile.lock:201]: In https://github.com/puppetlabs/vmpooler-provider-vsphere/pull/29/files#diff-89cade48462044ee1b672dc5f4c3ec250fbd29effcd8932096a23c1283c6731fL175 you remove one of the platforms... do these need to be consistent across repos?

**Review by genebean — APPROVED** [review]

File: .jrubyrc
Comment (yachub) [reply .jrubyrc:1]: I'll have to look into what that does. I generated this gem with `bundle gem`, so maybe a newer version or something.

File: Gemfile.lock (hunk 10 above)
Comment (yachub) [reply Gemfile.lock:201]: That must be leftover from when I started with jruby 9.3 then updated to 9.4 and didn't remove the old, so I'll remove that.

File: .rubocop.yml
Comment (yachub) [reply .rubocop.yml:1]: Oh darn, when I "fixed" that the first time I didn't notice the indentation on that list line.

**Review by isaac-hammes — APPROVED** [review]

**Pull request merged**